File: ios/bitchat/bitchat/Protocols/Geohash.swift
import Foundation

enum Geohash {
    private static let base32Chars = Array("0123456789bcdefghjkmnpqrstuvwxyz")
    private static let base32Map: [Character: Int] = {
        var map: [Character: Int] = [:]
        for (i, c) in base32Chars.enumerated() { map[c] = i }
        return map
    }()

    static func isValidBuildingGeohash(_ geohash: String) -> Bool {
        guard geohash.count == 8 else { return false }
        return geohash.lowercased().allSatisfy { base32Map[$0] != nil }
    }

    static func encode(latitude: Double, longitude: Double, precision: Int) -> String {
        guard precision > 0 else { return "" }

        var latInterval: (Double, Double) = (-90.0, 90.0)
        var lonInterval: (Double, Double) = (-180.0, 180.0)

        var isEven = true
        var bit = 0
        var ch = 0
        var geohash: [Character] = []

        let lat = max(-90.0, min(90.0, latitude))
        let lon = max(-180.0, min(180.0, longitude))

        while geohash.count < precision {
            if isEven {
                let mid = (lonInterval.0 + lonInterval.1) / 2
                if lon >= mid {
                    ch |= (1 << (4 - bit))
                    lonInterval.0 = mid
                } else {
                    lonInterval.1 = mid
                }
            } else {
                let mid = (latInterval.0 + latInterval.1) / 2
                if lat >= mid {
                    ch |= (1 << (4 - bit))
                    latInterval.0 = mid
                } else {
                    latInterval.1 = mid
                }
            }

            isEven.toggle()
            if bit < 4 {
                bit += 1
            } else {
                geohash.append(base32Chars[ch])
                bit = 0
                ch = 0
            }
        }

        return String(geohash)
    }

    static func decodeCenter(_ geohash: String) -> (lat: Double, lon: Double) {
        var latInterval: (Double, Double) = (-90.0, 90.0)
        var lonInterval: (Double, Double) = (-180.0, 180.0)

        var isEven = true
        for ch in geohash.lowercased() {
            guard let cd = base32Map[ch] else { continue }
            for mask in [16, 8, 4, 2, 1] {
                if isEven {
                    let mid = (lonInterval.0 + lonInterval.1) / 2
                    if (cd & mask) != 0 { lonInterval.0 = mid } else { lonInterval.1 = mid }
                } else {
                    let mid = (latInterval.0 + latInterval.1) / 2
                    if (cd & mask) != 0 { latInterval.0 = mid } else { latInterval.1 = mid }
                }
                isEven.toggle()
            }
        }
        let lat = (latInterval.0 + latInterval.1) / 2
        let lon = (lonInterval.0 + lonInterval.1) / 2
        return (lat, lon)
    }

    static func decodeBounds(_ geohash: String) -> (latMin: Double, latMax: Double, lonMin: Double, lonMax: Double) {
        var latInterval: (Double, Double) = (-90.0, 90.0)
        var lonInterval: (Double, Double) = (-180.0, 180.0)

        var isEven = true
        for ch in geohash.lowercased() {
            guard let cd = base32Map[ch] else { continue }
            for mask in [16, 8, 4, 2, 1] {
                if isEven {
                    let mid = (lonInterval.0 + lonInterval.1) / 2
                    if (cd & mask) != 0 { lonInterval.0 = mid } else { lonInterval.1 = mid }
                } else {
                    let mid = (latInterval.0 + latInterval.1) / 2
                    if (cd & mask) != 0 { latInterval.0 = mid } else { latInterval.1 = mid }
                }
                isEven.toggle()
            }
        }
        return (latInterval.0, latInterval.1, lonInterval.0, lonInterval.1)
    }

    static func neighbors(of geohash: String) -> [String] {
        guard !geohash.isEmpty else { return [] }

        let precision = geohash.count
        let bounds = decodeBounds(geohash)
        let center = decodeCenter(geohash)

        let latHeight = bounds.latMax - bounds.latMin
        let lonWidth = bounds.lonMax - bounds.lonMin

        func wrapLongitude(_ lon: Double) -> Double {
            var wrapped = lon
            while wrapped > 180.0 { wrapped -= 360.0 }
            while wrapped < -180.0 { wrapped += 360.0 }
            return wrapped
        }

        func clampLatitude(_ lat: Double) -> Double {
            return max(-90.0, min(90.0, lat))
        }

        let neighbors: [(lat: Double, lon: Double)] = [
            (center.lat + latHeight, center.lon),
            (center.lat + latHeight, center.lon + lonWidth),
            (center.lat, center.lon + lonWidth),
            (center.lat - latHeight, center.lon + lonWidth),
            (center.lat - latHeight, center.lon),
            (center.lat - latHeight, center.lon - lonWidth),
            (center.lat, center.lon - lonWidth),
            (center.lat + latHeight, center.lon - lonWidth)
        ]

        return neighbors.compactMap { neighbor in
            let lat = clampLatitude(neighbor.lat)
            let lon = wrapLongitude(neighbor.lon)

            if (neighbor.lat > 90.0 || neighbor.lat < -90.0) {
                return nil
            }

            return encode(latitude: lat, longitude: lon, precision: precision)
        }
    }
}

File: app/src/main/java/com/cybersiren/android/geohash/Geohash.kt
package com.cybersiren.android.geohash

object Geohash {

    private val base32Chars = "0123456789bcdefghjkmnpqrstuvwxyz".toCharArray()
    private val charToValue: Map<Char, Int> = base32Chars.withIndex().associate { it.value to it.index }

    data class Bounds(val latMin: Double, val latMax: Double, val lonMin: Double, val lonMax: Double)

    fun encode(latitude: Double, longitude: Double, precision: Int): String {
        if (precision <= 0) return ""

        var latInterval = -90.0 to 90.0
        var lonInterval = -180.0 to 180.0

        var isEven = true
        var bit = 0
        var ch = 0
        val geohash = StringBuilder()

        val lat = latitude.coerceIn(-90.0, 90.0)
        val lon = longitude.coerceIn(-180.0, 180.0)

        while (geohash.length < precision) {
            if (isEven) {
                val mid = (lonInterval.first + lonInterval.second) / 2
                if (lon >= mid) {
                    ch = ch or (1 shl (4 - bit))
                    lonInterval = mid to lonInterval.second
                } else {
                    lonInterval = lonInterval.first to mid
                }
            } else {
                val mid = (latInterval.first + latInterval.second) / 2
                if (lat >= mid) {
                    ch = ch or (1 shl (4 - bit))
                    latInterval = mid to latInterval.second
                } else {
                    latInterval = latInterval.first to mid
                }
            }

            isEven = !isEven
            if (bit < 4) {
                bit += 1
            } else {
                geohash.append(base32Chars[ch])
                bit = 0
                ch = 0
            }
        }

        return geohash.toString()
    }

    fun decodeToCenter(geohash: String): Pair<Double, Double> {
        val b = decodeToBounds(geohash)
        val latCenter = (b.latMin + b.latMax) / 2
        val lonCenter = (b.lonMin + b.lonMax) / 2
        return latCenter to lonCenter
    }

    fun decodeToBounds(geohash: String): Bounds {
        if (geohash.isEmpty()) return Bounds(0.0, 0.0, 0.0, 0.0)

        var latInterval = -90.0 to 90.0
        var lonInterval = -180.0 to 180.0
        var isEven = true

        geohash.lowercase().forEach { ch ->
            val cd = charToValue[ch] ?: return Bounds(0.0, 0.0, 0.0, 0.0)
            for (mask in intArrayOf(16, 8, 4, 2, 1)) {
                if (isEven) {
                    val mid = (lonInterval.first + lonInterval.second) / 2
                    if ((cd and mask) != 0) {
                        lonInterval = mid to lonInterval.second
                    } else {
                        lonInterval = lonInterval.first to mid
                    }
                } else {
                    val mid = (latInterval.first + latInterval.second) / 2
                    if ((cd and mask) != 0) {
                        latInterval = mid to latInterval.second
                    } else {
                        latInterval = latInterval.first to mid
                    }
                }
                isEven = !isEven
            }
        }
        return Bounds(
            latMin = minOf(latInterval.first, latInterval.second),
            latMax = maxOf(latInterval.first, latInterval.second),
            lonMin = minOf(lonInterval.first, lonInterval.second),
            lonMax = maxOf(lonInterval.first, lonInterval.second)
        )
    }

    fun neighborsSamePrecision(geohash: String): Set<String> {
        if (geohash.isEmpty()) return emptySet()
        val p = geohash.length
        val b = decodeToBounds(geohash)
        val dLat = b.latMax - b.latMin
        val dLon = b.lonMax - b.lonMin

        fun wrapLon(lon: Double): Double {
            var x = lon
            while (x > 180.0) x -= 360.0
            while (x < -180.0) x += 360.0
            return x
        }

        val neighbors = mutableSetOf<String>()
        for (dy in -1..1) {
            for (dx in -1..1) {
                if (dx == 0 && dy == 0) continue
                val centerLat = (b.latMin + b.latMax) / 2 + dy * dLat
                val rawLonCenter = (b.lonMin + b.lonMax) / 2 + dx * dLon
                val centerLon = wrapLon(rawLonCenter)
                val enc = encode(centerLat.coerceIn(-90.0, 90.0), centerLon, p)
                if (enc.isNotEmpty() && enc != geohash) neighbors.add(enc)
            }
        }
        return neighbors
    }
}
